File: Stage24/Python/Problem10830.py
import sys
input = sys.stdin.readline
MOD = 1000

def multiply(m1, m2, N):
    res = [[0] * N for _ in range(N)]
    for i in range(N):
        for j in range(N):
            for k in range(N):
                res[i][j] += m1[i][k] * m2[k][j]
            res[i][j] %= MOD
    return res

def power(matrix, exp, N):
    if exp == 1:
        return matrix

    half = power(matrix, exp // 2, N)
    result = multiply(half, half, N)

    if exp % 2 == 1:
        result = multiply(result, matrix, N)

    return result

def main():
    N, B = map(int, input().split())
    A = [list(map(lambda x: int(x) % MOD, input().split())) for _ in range(N)]
    
    result = power(A, B, N)

    for row in result:
        print(*row)
    
if __name__ == "__main__":
    main()

File: Stage24/C/Problem10830.c
#include <stdio.h>

#define MOD 1000
int N;
long long B;

void multiply(int a[5][5], int b[5][5], int result[5][5]) {
    int temp[5][5] = {0};

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            for (int k = 0; k < N; k++) {
                temp[i][j] += a[i][k] * b[k][j];
            }
            temp[i][j] %= MOD;
        }
    }

    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            result[i][j] = temp[i][j];
}

void power(int matrix[5][5], long long exp, int result[5][5]) {
    if (exp == 1) {
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                result[i][j] = matrix[i][j];
        return;
    }

    int half[5][5];
    power(matrix, exp / 2, half);

    int temp[5][5];
    multiply(half, half, temp);

    if (exp % 2 == 1) {
        multiply(temp, matrix, result);
    } else {
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                result[i][j] = temp[i][j];
    }
}

int main(void) {
    int A[5][5];
    int result[5][5];

    scanf("%d %lld", &N, &B);

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            scanf("%d", &A[i][j]);
            A[i][j] %= MOD;
        }
    }

    power(A, B, result);

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            printf("%d ", result[i][j]);
        }
        printf("\n");
    }

    return 0;
}